File: Leetcode_solutions/Move-Zeroes.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var moveZeroes = function(nums) {
  zeros = [];
  for (let i = 0; i < nums.length; i++) {
    if (copy[i] === 0) {
      zeros.push(copy[i]);
      nums.splice(i, 1);
      i--;
    }
  }
  for (let i = 0; i < zeros.length; i++) {
    nums.push(zeros[i]);
  }

  return nums;
};
console.log(moveZeroes([0, 1, 0, 3, 12]));
console.log(moveZeroes([0, 0, 1]));
console.log(moveZeroes([0, 0, 0, 1]));
console.log(moveZeroes([0, 1, 0, 3, 0, 0, 0, 12]));

File: Leetcode_solutions/moveZeroes.py
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        for i in range(len(nums) - 1, -1, -1):
            if nums[i] == 0:
                zero = nums[i]
                nums[i:] = nums[i + 1:]
                nums.append(zero)
        return nums

class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        lastNonZeroFoundAt = 0
        for i in range(len(nums)):
            if nums[i] != 0:
                nums[lastNonZeroFoundAt] = nums[i]
                lastNonZeroFoundAt += 1
        j = lastNonZeroFoundAt
        for j in range(lastNonZeroFoundAt, len(nums)):
            nums[j] = 0
            

class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        lastNonZeroFoundAt = 0
        for cur in range(len(nums)):
            if nums[cur] != 0:
                nums[lastNonZeroFoundAt], nums[cur] = nums[cur], nums[lastNonZeroFoundAt]
                lastNonZeroFoundAt += 1